contracts/chain-adapters/Solana_Adapter.sol

```diff
@@ -0,0 +1,193 @@
+    address public immutable SOLANA_USDC_ADDRESS;
+
+    // USDC token address on Solana for the spoke pool (vault ATA), decoded from Base58 to bytes32.
+    bytes32 public immutable SOLANA_SPOKE_POOL_USDC_VAULT;
+
+    // Custom errors for constructor argument validation.
+    error InvalidCctpTokenMessenger(address tokenMessenger);
+    error InvalidCctpMessageTransmitter(address messageTransmitter);
+
+    // Custom errors for relayMessage validation.
+    error InvalidRelayMessageTarget(address target);
+    error InvalidOriginToken(address originToken);
+    error InvalidDestinationChainId(uint256 destinationChainId);
+
+    // Custom errors for relayTokens validation.
+    error InvalidL1Token(address l1Token);
+    error InvalidL2Token(address l2Token);
+    error InvalidAmount(uint256 amount);
+    error InvalidTokenRecipient(address to);
+
+    /**
+     * @notice Constructs new Adapter.
+     * @param _l1Usdc USDC address on L1.
+     * @param _cctpTokenMessenger TokenMessenger contract to bridge tokens via CCTP.
+     * @param _cctpMessageTransmitter MessageTransmitter contract to bridge messages via CCTP.
+     * @param solanaSpokePool Solana spoke pool address, decoded from Base58 to bytes32.
+     * @param solanaUsdc USDC mint address on Solana, decoded from Base58 to bytes32.
+     * @param solanaSpokePoolUsdcVault USDC token address on Solana for the spoke pool, decoded from Base58 to bytes32.
+     */
+    constructor(
+        IERC20 _l1Usdc,
+        ITokenMessenger _cctpTokenMessenger,
+        IMessageTransmitter _cctpMessageTransmitter,
+        bytes32 solanaSpokePool,
+        bytes32 solanaUsdc,
+        bytes32 solanaSpokePoolUsdcVault
+    ) CircleCCTPAdapter(_l1Usdc, _cctpTokenMessenger, CircleDomainIds.Solana) {
+        // Solana adapter requires CCTP TokenMessenger and MessageTransmitter contracts to be set.
+        if (address(_cctpTokenMessenger) == address(0)) {
+            revert InvalidCctpTokenMessenger(address(_cctpTokenMessenger));
+        }
+        if (address(_cctpMessageTransmitter) == address(0)) {
+            revert InvalidCctpMessageTransmitter(address(_cctpMessageTransmitter));
+        }
+
+        cctpMessageTransmitter = _cctpMessageTransmitter;
+
+        SOLANA_SPOKE_POOL_BYTES32 = solanaSpokePool;
+        SOLANA_SPOKE_POOL_ADDRESS = _mapSolanaAddress(solanaSpokePool);
+
+        SOLANA_USDC_BYTES32 = solanaUsdc;
+        SOLANA_USDC_ADDRESS = _mapSolanaAddress(solanaUsdc);
+
+        SOLANA_SPOKE_POOL_USDC_VAULT = solanaSpokePoolUsdcVault;
+    }
+
+    /**
+     * @notice Send cross-chain message to target on Solana.
+     * @dev Only allows sending messages to the Solana spoke pool.
+     * @param target Program on Solana (translated as EVM address) that will receive message.
+     * @param message Data to send to target.
+     */
+    function relayMessage(address target, bytes calldata message) external payable override {
+        if (target != SOLANA_SPOKE_POOL_ADDRESS) {
```

Comment:
what is the downside with letting this target non spoke pool address?

Reply:
Having it hardcoded potentially limits flexibility for future updates or use cases. Maybe we can add a setter or remove the check if not strictly required as @chrismaree is exploring.

Reply:
This accepts target address that must be translated to bytes32, but since we don't have translation mapping for any other target then obviously the caller has been misconfigured.

As for updates, I understand the general pattern for these adapters is to have them immutable and upgrade by redeploying and updating the instance in HubPool.

Reply:
@Reinis-FRP that's a good point. this is _not_ a generic adapter as a result of the encoding so there is no situation where you should ever be passing in anything other than a known target address. if you are, you're doing something wrong and it should error. I agree with leaving it like this.